doc/development/guide/documentation.rst

```diff
@@ -1,39 +1,449 @@
+are automatically extracted and rendered by Sphinx, and inserted into the
+:doc:`API </code/qml>` section of the documentation.
+
+Guiding principles
+~~~~~~~~~~~~~~~~~~
+
+Below, we list some general guidelines, before discussing required docstring styles and formatting
+for various PennyLane objects.
+
+* **Be clear and concise, and avoid redundant information.** Favour clarity over completeness; too
+  much information can distract from conveying the functionality of the code unit.
+  Assume the reader has basic domain knowledge, and provide links to internal/external resources
+  where deemed necessary.
+
+* **Be mindful with ReST markup**. Overly complex formatting can make the content difficult
+  to read for users viewing the docstring in a terminal/Jupyter environment. Try to also avoid
+  complex formatting such as tables, which can be difficult to maintain over time.
+
+* **Do not assume the reader can view the accompanying code.** The reader will most likely
+  be viewing the docstring via the Sphinx-rendered website, or in a terminal/Jupyter environment.
+  For documenting how the code itself works, favour comments.
+
+* **Cross reference where applicable.** Use the roles ``:class:``, ``:func:``, ``:attr:``,
+  ``:meth:``, and ``:mod:`` where applicable to have Sphinx automatically include a cross-reference
+  to the specified code object.
+
+  For example:
+
+  .. code-block:: rest
+
+      The recommended method of creating a QNodeCollection is via :func:`~.map`.
+
+  Here, ``~`` instructs Sphinx to display just the object name, as opposed to the full absolute path
+  of the object (``pennylane.map()``). To use custom text, you may use the syntax
+  ``:func:`map function <~.map>```.
+
+* **Favour document links over reference links.** If linking to a particular page, use
+  ``:doc:`relative/path/filename``` or ``:doc:`/absolute/path/filename``` rather than ``:ref:``.
+
+
+Modules
+~~~~~~~
+
+Every module should begin with the license boilerplate, followed by a module docstring.
+
+The module docstring must contain a short, single sentence summary of the module, followed
+by an (optional) multi-line description of the contents and usage.
+
+.. code-block:: rest
+
+    """
+    Single sentence that summarizes the module and its contents.
+
+    Optional multi-line description. Make sure to focus on functionality first and
+    foremost, and implementation details or background theory afterwards if required.
+    """
+
+Do not manually list the contents of the module; PennyLane uses
+`sphinx-automodapi <https://github.com/astropy/sphinx-automodapi>`__ to automatically
+create module listings.
+
+
+Functions and methods
+~~~~~~~~~~~~~~~~~~~~~
+
+The docstring should provide enough information for the reader to call and use the
+function or method without having access tot he code.
+Most functions or methods should have the following structure:
+
+.. code-block:: python
+
+    def func(arg1, arg2, **kwargs):
```

Comment:
```suggestion
    def func(arg1, arg2, kwarg1=val1):
```

Reply:
Oh, `**kwargs` was intentional here. Named arguments with default values we always include under `Args:`.

Reply:
k, be good to make that clearer

Reply:
Done